File: src/main/java/com/gregtechceu/gtceu/api/sync_system/codecs/CustomFluidTankCodec.java
package com.gregtechceu.gtceu.api.sync_system.codecs;

import com.gregtechceu.gtceu.api.sync_system.ContextualFieldCodec;
import com.gregtechceu.gtceu.api.transfer.fluid.CustomFluidTank;
import com.gregtechceu.gtceu.common.data.GTDataComponents;
import com.gregtechceu.gtceu.common.data.datacomponents.TransferData;

import net.minecraft.core.component.DataComponentMap;

import com.google.gson.JsonElement;
import com.mojang.serialization.JsonOps;

public final class CustomFluidTankCodec implements ContextualFieldCodec<CustomFluidTank> {

    public static final Class<CustomFluidTank> TYPE = CustomFluidTank.class;
    public static final CustomFluidTankCodec INSTANCE = new CustomFluidTankCodec();

    private CustomFluidTankCodec() {}

    @Override
    public JsonElement serializeField(CustomFluidTank value, Context<CustomFluidTank> context) {
        return DataComponentMap.CODEC
                .encodeStart(context.lookup().createSerializationContext(JsonOps.INSTANCE), value.exportComponents())
                .getOrThrow();
    }

    @Override
    public CustomFluidTank deserializeField(JsonElement value, Context<CustomFluidTank> context) {
        DataComponentMap components = DataComponentMap.CODEC
                .parse(context.lookup().createSerializationContext(JsonOps.INSTANCE), value)
                .getOrThrow();
        TransferData.FluidTank data = components.get(GTDataComponents.TRANSFER_FLUID_TANK.get());
        CustomFluidTank tank = context.currentValue();
        if (data == null && tank == null) {
            throw new IllegalArgumentException("Sync: fluid tank field " + context.fieldName() +
                    " is missing transfer_fluid_tank");
        }
        if (tank == null) {
            tank = new CustomFluidTank(data.capacity());
        }

        tank.importComponents(components);
        return tank;
    }
}

File: src/main/java/com/gregtechceu/gtceu/api/sync_system/codecs/CustomFluidTankCodec.kt
package com.gregtechceu.gtceu.api.sync_system.codecs

import com.gregtechceu.gtceu.api.sync_system.ContextualFieldCodec
import com.gregtechceu.gtceu.api.transfer.fluid.CustomFluidTank
import com.gregtechceu.gtceu.common.data.GTDataComponents

import net.minecraft.core.component.DataComponentMap

import com.google.gson.JsonElement
import com.mojang.serialization.JsonOps

class CustomFluidTankCodec private constructor() : ContextualFieldCodec<CustomFluidTank> {

	override fun serializeField(value: CustomFluidTank, context: ContextualFieldCodec.Context<CustomFluidTank>): JsonElement = DataComponentMap.CODEC
		.encodeStart(context.lookup.createSerializationContext(JsonOps.INSTANCE), value.exportComponents())
		.getOrThrow()

	override fun deserializeField(value: JsonElement, context: ContextualFieldCodec.Context<CustomFluidTank>): CustomFluidTank {
		val components = DataComponentMap.CODEC
			.parse(context.lookup.createSerializationContext(JsonOps.INSTANCE), value)
			.getOrThrow()
		val data = components.get(GTDataComponents.TRANSFER_FLUID_TANK.get())
		val tank = context.currentValue ?: CustomFluidTank(
			data?.capacity ?: throw IllegalArgumentException("Sync: fluid tank field ${context.fieldName} is missing transfer_fluid_tank"),
		)
		tank.importComponents(components)
		return tank
	}

	companion object {
		@JvmField
		val TYPE: Class<CustomFluidTank> = CustomFluidTank::class.java

		@JvmField
		val INSTANCE = CustomFluidTankCodec()
	}
}
